pkg/psalabelsyncer/sccrolecache.go

```diff
@@ -0,0 +1,440 @@
+}
+
+// getRoleFromRoleRef tries to retrieve the role or clusterrole from roleRef.
+func (c *SAToSCCCache) getRoleFromRoleRef(ns string, roleRef rbacv1.RoleRef) (*roleObj, error) {
+	var role interface{}
+	var retryErr error
+	switch kind := roleRef.Kind; kind {
+	case "Role":
+		retryErr = retry.OnError(retry.DefaultBackoff, apierrors.IsNotFound, func() error {
+			var err error
+			role, err = c.roleLister.Roles(ns).Get(roleRef.Name)
+			return err
+		})
+
+	case "ClusterRole":
+		retryErr = retry.OnError(retry.DefaultBackoff, apierrors.IsNotFound, func() error {
+			var err error
+			role, err = c.clusterRoleLister.Get(roleRef.Name)
+			return err
+		})
+
+	default:
+		return nil, fmt.Errorf("unknown kind in roleRef: %s", kind)
+	}
+
+	if retryErr != nil {
+		return nil, retryErr
+	}
+
+	return newRoleObj(role)
+}
+
+// IsRoleBindingRelevant returns true if the cluster/rolebinding supplied binds
+// to a Role that provides access to at least one of the SCCs available in the
+// cluster.
+func (c *SAToSCCCache) IsRoleBindingRelevant(obj interface{}) bool {
+	rb, err := newRoleBindingObj(obj)
+	if err != nil {
+		klog.Warningf("unexpected error, this may be a bug: %v", err)
+		return false
+	}
+
+	role, err := c.getRoleFromRoleRef(rb.Namespace(), rb.RoleRef())
+	if err != nil {
+		klog.Infof("failed to retrieve a role for a rolebinding ref: %v", err)
+		return false
+	}
+
+	return c.IsRoleInvolvesSCCs(role, false)
+}
+
+// IsRoleInvolvesSCCs returns true if the role supplied in obj provides access
+// to at least one of the SCCs present in the cluster.
+// Set isRoleUpdate to true if instead of using the cached role the supplied object
+// should be used to update the role in the cache as well.
+func (c *SAToSCCCache) IsRoleInvolvesSCCs(obj interface{}, isRoleUpdate bool) bool {
+	// lock the roles cache to make sure the state of the role that's later retrieved here
+	// is written properly
+	c.usefulRolesLock.Lock()
+	defer c.usefulRolesLock.Unlock()
+
+	role, err := newRoleObj(obj)
+	if err != nil {
+		klog.Warningf("unexpected error, this may be a bug: %v", err)
+		return false
+	}
+
+	if isRoleUpdate {
+		sccs, err := c.sccLister.List(labels.Everything())
+		if err != nil {
+			klog.Warning("failed to list SCCs: %v", err)
+			return false
+		}
+
+		c.syncRoleCache(role.Namespace(), role.Name(), role.Rules(), sccs)
+	}
+
+	return len(c.usefulRoles[fmt.Sprintf("%s/%s", role.Namespace(), role.Name())]) != 0
+}
+
+// ReinitializeRoleCache clears the current cache of roles that provide access
+// to SCCs and reinitializes it by pulling all cluster-/roles and SCCs and
+// reevaluating them.
+// This should be used rather rarely as there are many computations involved
+// when assessing all the present cluster-/role rules.
+func (c *SAToSCCCache) ReinitializeRoleCache() error {
+	// rewriting the whole cache, lock is needed
+	c.usefulRolesLock.Lock()
+	defer c.usefulRolesLock.Unlock()
+
+	c.usefulRoles = map[string][]string{}
```

Comment:
I'm not seeing where this is updated over time as roles are updated.

Reply:
Hmm, previously I had it tied to events that come for the controller but it's true that while thinking of it as a separate logical object, it should add its own event handlers to the informers.

Reply:
I implemented it so that the informer events for roles and SCCs are handled by the cache, and the cache is capable of adding keys to an external queue.